docs/examples/advanced_optimizations.ipynb

```diff
@@ -0,0 +1,330 @@
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 5,
+   "id": "a7f612ec",
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Mean time: 27.510029296875 ms\n"
+     ]
+    }
+   ],
+   "source": [
+    "# Construct layer\n",
+    "wgrad_transformer = te.TransformerLayer(\n",
+    "    hidden_size,\n",
+    "    ffn_hidden_size,\n",
+    "    num_attention_heads,\n",
+    "    fuse_wgrad_accumulation=True,\n",
+    "    fuse_qkv_params=True, # Required for fuse_wgrad_accumulation\n",
+    ")\n",
+    "wgrad_transformer.to(dtype=dtype).cuda()\n",
+    "for param in wgrad_transformer.parameters():\n",
+    "    param.grad = None\n",
+    "    param.main_grad = torch.zeros_like(param, dtype=torch.float32)\n",
+    "\n",
+    "# Training step\n",
+    "with te.fp8_autocast(enabled=True, fp8_recipe=fp8_recipe):\n",
+    "    y = wgrad_transformer(x, attention_mask=None)\n",
+    "y.backward(dy)\n",
+    "for param in wgrad_transformer.parameters():\n",
+    "    if param.grad is not None:\n",
+    "        param.main_grad.copy_(param.grad)\n",
+    "        param.grad = None\n",
+    "\n",
+    "# Measure step time\n",
+    "utils.speedometer(\n",
+    "    wgrad_transformer,\n",
+    "    x,\n",
+    "    dy,\n",
+    "    forward_kwargs = { \"attention_mask\": None },\n",
+    "    fp8_autocast_kwargs = { \"enabled\": True, \"fp8_recipe\": fp8_recipe },\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "add64bd5",
+   "metadata": {},
+   "source": [
+    "## FP8 weight caching\n",
+    "\n",
+    "<div class=\"alert alert-info\">\n",
+    "\n",
+    "<b>Summary</b>\n",
+    "    \n",
+    "We avoid redundant FP8 casting when training with multiple gradient accumulation steps.\n",
+    "\n",
+    "</div>\n",
+    "\n",
+    "Since weights are typically trained in FP32, a type conversion is required before we can perform compute in FP8. By default, the [fp8_autocast](../api/pytorch.rst#transformer_engine.pytorch.fp8_autocast) context manager will handle this internally by casting non-FP8 tensors to FP8 as they are encountered. However, we can improve upon this if we assume we are following a standard training loop. In particular, if we split up a training iteration into multiple gradient accumulation steps, each micro-batch will encounter the same weight tensors. Thus, we only need to cast the weights to FP8 in the first gradient accumulation step and we can cache the FP8 weights for the remaining gradient accumulation steps."
```

Comment:
```suggestion
    "Since weights are typically trained in FP32, a type conversion is required before we can perform compute in FP8. By default, the [fp8_autocast](../api/pytorch.rst#transformer_engine.pytorch.fp8_autocast) context manager will handle this internally by casting non-FP8 tensors to FP8 as they are encountered. However, we can improve upon this in some cases. In particular, if our training iteration is split into multiple gradient accumulation steps, each micro-batch will encounter the same weight tensors. Thus, we only need to cast the weights to FP8 in the first gradient accumulation step and we can cache the resulting FP8 weights for the remaining gradient accumulation steps."
```